### PR TITLE
Fixed #30360 -- Support rotation of secret keys

### PR DESCRIPTION
**Pull request by pelme**
[Ticket](https://code.djangoproject.com/ticket/30360) & [django-dev discussion](https://groups.google.com/forum/#!msg/django-developers/jg-eu3HtLHI/V_rbzYKfAQAJ)

This PR is work-in-progress. I intend to discuss and work on it during the Djangocon 2019 sprints.å

### CODE REVIEW

File: django/core/secret_key.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+
+class SecretKeySettingsBackend:
+    def get_secret_key(self):
+        return settings.SECRET_KEY
+
+    def get_old_keys(self):
+        return list(settings.OLD_SECRET_KEYS)
+
+    def check_secret_key(self):
```
Comment (adamchainz) [inline django/core/secret_key.py]: `check` would be a more conventional name

File: django/core/secret_key.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+        passed_check = (
+            getattr(settings, 'SECRET_KEY', None) and
+            len(set(settings.SECRET_KEY)) >= SECRET_KEY_MIN_UNIQUE_CHARACTERS and
+            len(settings.SECRET_KEY) >= SECRET_KEY_MIN_LENGTH
+        )
+        return [] if passed_check else [W009]
+
+
+def _get_backend():
+    return import_string(settings.SECRET_KEY_BACKEND)()
```
Comment (adamchainz) [inline django/core/secret_key.py]: This will instantiate the backend every time, it should be instantiated only once, and re-built when the setting changes. The other backends should have clues.

File: django/core/secret_key.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
+from django.conf import settings
```
Comment (adamchainz) [inline django/core/secret_key.py]: `django.core.secret_keys` ?

File: django/core/secret_key.py [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+def get_secret_key():
+    return _get_backend().get_secret_key()
+
+
+def get_old_keys():
+    return _get_backend().get_old_keys()
+
+
+def check_secret_key():
```
Comment (adamchainz) [inline django/core/secret_key.py]: `check`

File: django/core/signing.py [hunk 5]
```diff
@@ -90,10 +101,17 @@ def loads(self, data):
         return json.loads(data.decode('latin-1'))
 
 
-def dumps(obj, key=None, salt='django.core.signing', serializer=JSONSerializer, compress=False):
+def dumps(
+    obj,
+    key=None,
+    old_keys=None,
```
Comment (adamchainz) [inline django/core/signing.py]: the old keys aren't needed when dumping so I don't think this function needs to change

File: django/core/signing.py [hunk 6]
```diff
@@ -120,7 +138,7 @@ def dumps(obj, key=None, salt='django.core.signing', serializer=JSONSerializer,
     base64d = b64_encode(data).decode()
     if is_compressed:
         base64d = '.' + base64d
-    return TimestampSigner(key, salt=salt).sign(base64d)
+    return TimestampSigner(key, old_keys=old_keys, salt=salt).sign(base64d)
 
 
 def loads(s, key=None, salt='django.core.signing', serializer=JSONSerializer, max_age=None):
```
Comment (adamchainz) [inline django/core/signing.py]: This would need multiple keys so should have an extra argument?

File: django/core/signing.py [hunk 7]
```diff
@@ -144,9 +162,19 @@ def loads(s, key=None, salt='django.core.signing', serializer=JSONSerializer, ma
...
-    def __init__(self, key=None, sep=':', salt=None):
-        # Use of native strings in all versions of Python
-        self.key = key or settings.SECRET_KEY
+    def __init__(self, key=None, sep=':', old_keys=None, salt=None):
+        if key is None:
+            if old_keys is not None:
+                raise ImproperlyConfigured('You must specify key when you specify old_keys.')
+
+            self.key = get_secret_key()
+            self.old_keys = [self.key] + get_old_keys()
```
Comment (adamchainz) [inline django/core/signing.py]: `all_keys` ?

File: django/core/secret_key.py [hunk 8]
```diff
@@ -0,0 +1,39 @@
+from django.conf import settings
+from django.utils.module_loading import import_string
+
+
+class SecretKeySettingsBackend:
```
Comment (adamchainz) [inline django/core/secret_key.py]: `SettingsSecretKeys` a better name, other backends can be called e.g. `DatabaseSecretKeys` ?

File: django/conf/global_settings.py [hunk 9]
```diff
@@ -263,6 +263,12 @@ def gettext_noop(s):
 # loudly.
 SECRET_KEY = ''
 
+# A list of secret keys that can be used in addition to the SECRET_KEY to
+# verify signatures.
+OLD_SECRET_KEYS = []
+
+SECRET_KEY_BACKEND = 'django.core.secret_key.SecretKeySettingsBackend'
```
Comment (adamchainz) [inline django/conf/global_settings.py]: `SECRET_KEYS_BACKEND` probably better

File: django/core/secret_key.py [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+        from django.core.checks.security.base import (
+            SECRET_KEY_MIN_LENGTH, SECRET_KEY_MIN_UNIQUE_CHARACTERS, W009,
+        )
+
+        passed_check = (
+            getattr(settings, 'SECRET_KEY', None) and
+            len(set(settings.SECRET_KEY)) >= SECRET_KEY_MIN_UNIQUE_CHARACTERS and
+            len(settings.SECRET_KEY) >= SECRET_KEY_MIN_LENGTH
+        )
+        return [] if passed_check else [W009]
```
Comment (adamchainz) [inline django/core/secret_key.py]: what about checking the old keys?

I think all the check warning and constants should be moved in here too

File: django/core/checks/security/base.py [hunk 11]
```diff
@@ -182,12 +182,8 @@ def check_ssl_redirect(app_configs, **kwargs):
 
 @register(Tags.security, deploy=True)
 def check_secret_key(app_configs, **kwargs):
```
Comment (adamchainz) [inline django/core/checks/security/base.py]: `check_secret_keys`

File: django/core/secret_key.py (hunk 3 above)
Comment (pelme) [reply django/core/secret_key.py]: Not relevant anymore since this PR no longer contains the settings backend stuff. :)

File: django/core/secret_key.py (hunk 8 above)
Comment (pelme) [reply django/core/secret_key.py]: Not relevant anymore since this PR no longer contains the settings backend stuff. :)

File: django/core/secret_key.py (hunk 10 above)
Comment (pelme) [reply django/core/secret_key.py]: Done, now all keys are checked.

File: django/core/secret_key.py (hunk 2 above)
Comment (pelme) [reply django/core/secret_key.py]: Not relevant anymore since this PR no longer contains the settings backend stuff. :)

File: django/core/signing.py (hunk 7 above)
Comment (pelme) [reply django/core/signing.py]: This has changed in the latest revision and is handled in another way now.

File: django/core/secret_key.py (hunk 4 above)
Comment (pelme) [reply django/core/secret_key.py]: Not relevant anymore since this PR no longer contains the settings backend stuff. :)

File: django/conf/global_settings.py (hunk 9 above)
Comment (pelme) [reply django/conf/global_settings.py]: Not relevant anymore since this PR no longer contains the settings backend stuff. :)

**Comment by pelme** [comment]
@MarkusH @adamchainz I have reworked this PR now, getting rid of the settings backend stuff and focusing on having SECRET_KEY and SECRET_KEYS as discussed with Markus.

Docs and release notes are still missing, would be happy if you could have a quick look before I start working writing them up to ensure this is going in the right direction. :)

File: django/conf/__init__.py [hunk 12]
```diff
@@ -157,8 +157,15 @@ def __init__(self, settings_module):
                 setattr(self, setting, setting_value)
                 self._explicit_settings.add(setting)
 
+        if self.SECRET_KEY:
+            if self.SECRET_KEYS is not None:
+                raise ImproperlyConfigured('Only one of SECRET_KEY and SECRET_KEYS must be specified, not both.')
```
Comment (MarkusH) [inline django/conf/__init__.py]:
```suggestion
                raise ImproperlyConfigured('Only one of SECRET_KEY and SECRET_KEYS can be specified, not both.')
```

File: django/core/signing.py [hunk 13]
```diff
@@ -72,10 +74,16 @@ def base64_hmac(salt, value, key):
     return b64_encode(salted_hmac(salt, value, key).digest()).decode()
 
 
+def _cookie_signer_key(key):
+    return b'django.http.cookies' + force_bytes(key)
```
Comment (MarkusH) [inline django/core/signing.py]: Let's carry over the comment.

```suggestion
    return b'django.http.cookies' + force_bytes(key)  # SECRET_KEY may be str or bytes.
```

File: django/core/signing.py [hunk 14]
```diff
@@ -93,7 +101,7 @@ def loads(self, data):
 def dumps(obj, key=None, salt='django.core.signing', serializer=JSONSerializer, compress=False):
     """
     Return URL-safe, hmac/SHA1 signed base64 compressed JSON string. If key is
-    None, use settings.SECRET_KEY instead.
+    None, use the default secret key instead.
```
Comment (MarkusH) [inline django/core/signing.py]:
```suggestion
    None, use the first key in settings.SECRET_KEYS instead.
```

File: django/core/signing.py [hunk 15]
```diff
@@ -123,15 +131,15 @@ def dumps(obj, key=None, salt='django.core.signing', serializer=JSONSerializer,
...
+def loads(s, key=None, salt='django.core.signing', serializer=JSONSerializer, max_age=None, keys=None):
     """
     Reverse of dumps(), raise BadSignature if signature fails.
 
     The serializer is expected to accept a bytestring.
     """
     # TimestampSigner.unsign() returns str but base64 and zlib compression
     # operate on bytes.
-    base64d = TimestampSigner(key, salt=salt).unsign(s, max_age=max_age).encode()
+    base64d = TimestampSigner(key, salt=salt, keys=keys).unsign(s, max_age=max_age).encode()
```
Comment (MarkusH) [inline django/core/signing.py]: What do you think about making the call args a bit more explicit?

```suggestion
    base64d = TimestampSigner(key=key, salt=salt, keys=keys).unsign(s, max_age=max_age).encode()
```

File: tests/settings_tests/tests.py [hunk 16]
```diff
@@ -288,16 +288,6 @@ def test_override_settings_nested(self):
         with self.assertRaises(AttributeError):
             getattr(settings, 'TEST2')
 
-    def test_no_secret_key(self):
```
Comment (MarkusH) [inline tests/settings_tests/tests.py]: Why did you need to remove this test?

**Comment by adamchainz** [comment]
Hey @pelme , thanks for coming back to this. I see you just merged master into this branch. It'd be better to rebase on top of the latest master :)

File: tests/auth_tests/test_tokens.py [hunk 17]
```diff
@@ -69,9 +69,34 @@ def test_token_with_different_secret(self):
...
+        user = User.objects.create_user('tokentestuser', 'test2@example.com', 'testpw')
+
+        p1 = PasswordResetTokenGenerator()
+        p1.secret = 'secret'
+
+        p2 = PasswordResetTokenGenerator()
+        p2.secrets = ['newsecret', 'secret']
+
+        tk1 = p1.make_token(user)
+        self.assertTrue(p1.check_token(user, tk1))
```
Comment (atombrella) [inline tests/auth_tests/test_tokens.py:87]: Please use `assertIs` for boolean checks. https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/coding-style/#python-style (see under `assertIs`).

**Comment by carltongibson** [comment]
@terrameijar This was the PR we were talking about. Are you good to get it checked out? Ping me on the forum as you're getting set up, or when you are set up, and I can help you get going, and we can discuss.

**Comment by terrameijar** [comment]
@carltongibson thanks I'll have a go at setting up my environment tomorrow.
I'll definitely ping you for help. Thank you

**Comment by carltongibson** [comment]
No problem. Thank you! Welcome aboard, don't be shy to call out!

**Comment by felixxm** [comment]
Superseded by #13618.
